### PR TITLE
Validate for pack

### PR DESCRIPTION
**Pull request by Shellyber**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/22759

## Screenshots
image

## Must have
- [ ] Tests

### CODE REVIEW

**Comment by Shellyber** [comment]
Working on tests

File: demisto_sdk/__main__.py [hunk 1]
```diff
@@ -208,22 +208,31 @@ def unify(**kwargs):
     '-a', '--validate-all', is_flag=True, show_default=True, default=False,
     help='Whether to run all validation on all files or not'
 )
+@click.option(
+    '-i', '--input', help='The path of a pack to validate specifically.'
```
Comment (Itay4) [inline demisto_sdk/__main__.py:212]: i dont think we need another option for that, lets use `-p`, otherwise it will just be confusing

File: demisto_sdk/commands/validate/file_validator.py [hunk 2]
```diff
@@ -714,6 +733,9 @@ def is_valid_structure(self):
         if self.validate_all:
             self.validate_all_files()
             return self._is_valid
+        if self.pack_path:
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:736]: once you remove the additional flag, this should be moved into `if self.file_path:`

File: demisto_sdk/commands/validate/file_validator.py [hunk 3]
```diff
@@ -650,6 +651,24 @@ def validate_all_files(self):
         for file in all_files_to_validate:
             self.run_all_validations_on_file(file, file_type=find_type(file))
 
+    def validate_pack(self):
+        """Validate files in a specified pack"""
+        packs = glob(f'{PACKS_DIR}/*')
+        if self.pack_path not in packs:
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:657]: not sure this is needed, as you already check the directory exist in `__main__`

File: demisto_sdk/commands/validate/file_validator.py [hunk 4]
```diff
@@ -650,6 +651,24 @@ def validate_all_files(self):
...
+            raise Exception(F'File {self.pack_path} was not found\nMake sure you are using relative path')
+        pack_files = {file for file in glob(fr'{self.pack_path}/**', recursive=True) if not os.path.isdir(file)}
+        pack_unique_files_validator = PackUniqueFilesValidator(os.path.abspath(self.pack_path))
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:660]: just call `validate_pack_unique_files()` from this class instead, it does exactly that

File: demisto_sdk/commands/validate/file_validator.py [hunk 5]
```diff
@@ -650,6 +651,24 @@ def validate_all_files(self):
...
+        pack_errors = pack_unique_files_validator.validate_pack_unique_files()
+        if pack_errors:
+            print_error(pack_errors)
+            self._is_valid = False
+        for file in pack_files:
+            if file.endswith('pack_metadata.json'):
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:666]: this does not need to be validated because it is validated in `validate_pack_unique_files` ? what about the rest of the files validated there? why are we not skipping them ?

File: demisto_sdk/commands/validate/file_validator.py [hunk 6]
```diff
@@ -650,6 +651,24 @@ def validate_all_files(self):
...
+                continue
+            if any(test_file in file.lower() for test_file in TESTS_DIRECTORIES):
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:668]: this seem to be useful as tool func, like `is_test_file()`, what do you think?

File: demisto_sdk/commands/validate/file_validator.py [hunk 7]
```diff
@@ -650,6 +651,24 @@ def validate_all_files(self):
         for file in all_files_to_validate:
             self.run_all_validations_on_file(file, file_type=find_type(file))
 
+    def validate_pack(self):
+        """Validate files in a specified pack"""
+        packs = glob(f'{PACKS_DIR}/*')
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:656]: lets add a print before everything starts `Validating Pack <PACK-NAME>`

File: demisto_sdk/__main__.py (hunk 1 above)
Comment (Shellyber) [reply demisto_sdk/__main__.py:212]: A lot of our commands, using -i as input. It just to keep it consistent. 
Change anyway? 
Also the -p exists for specific file path

File: demisto_sdk/commands/validate/file_validator.py (hunk 2 above)
Comment (Shellyber) [reply demisto_sdk/commands/validate/file_validator.py:736]: What do you mean here?

File: demisto_sdk/commands/validate/file_validator.py (hunk 5 above)
Comment (Shellyber) [reply demisto_sdk/commands/validate/file_validator.py:666]: No need. The `run_all_validations_on_file` is not checking anything that is not "yml", "md" "json". so secrets and pack ignore skipped in the check anyways

Comment (Shellyber) [reply demisto_sdk/commands/validate/file_validator.py:666]: And yes, all pack unique files, are validated in `validate_pack_unique_file`

File: demisto_sdk/__main__.py (hunk 1 above)
Comment (Itay4) [reply demisto_sdk/__main__.py:212]: i just dont want to have both
i agree we should stick to `-i`, so lets set `-p` to `hidden=True` and `-i` for all kind of inputs: files, packs, etc

File: demisto_sdk/commands/validate/file_validator.py (hunk 5 above)
Comment (Itay4) [reply demisto_sdk/commands/validate/file_validator.py:666]: ok lets add a comment here for that

File: demisto_sdk/commands/validate/file_validator.py (hunk 2 above)
Comment (Itay4) [reply demisto_sdk/commands/validate/file_validator.py:736]: it you look at line `if self.file_path:`, thats where we actually use the user input of file to validate, so this should be moved there

File: demisto_sdk/commands/validate/file_validator.py (hunk 4 above)
Comment (Shellyber) [reply demisto_sdk/commands/validate/file_validator.py:660]: Yeah, but I want to give it a specific path to the pack. 
Else, it will go over all packs. That's why I did it this way

Comment (Itay4) [reply demisto_sdk/commands/validate/file_validator.py:660]: it will iterate all the packs you give it, so if you give one pack, which is the one you already got, it will run only on it
i would like to avoid this code duplication as it might cause us issues in the future maintaining it

File: demisto_sdk/__main__.py [hunk 8]
```diff
@@ -202,27 +202,30 @@ def unify(**kwargs):
...
 @pass_config
 def validate(config, **kwargs):
     sys.path.append(config.configuration.env_dir)
 
-    file_path = kwargs['path']
-
-    if file_path and not os.path.isfile(file_path):
+    file_path = kwargs['path'] or kwargs['input']
+    if file_path and not os.path.isfile(file_path) and not os.path.isdir(file_path):
         print_error(F'File {file_path} was not found')
```
Comment (Itay4) [inline demisto_sdk/__main__.py:225]:
```suggestion
        print_error(f'File {file_path} was not found')
```

File: demisto_sdk/commands/validate/README.md [hunk 9]
```diff
@@ -63,3 +65,6 @@ This will validate the file Integrations/Pwned-V2/Pwned-V2.yml only.
 `demisto-sdk validate -a`
 This will validate all files under `Packs` and `Beta_Integrations` directories
 <br><br>
+`demisto-sdk validate -i Packs/HelloWorld`
+This will validate all files under the `Pack`
```
Comment (Itay4) [inline demisto_sdk/commands/validate/README.md:70]:
```suggestion
This will validate all files under the content pack `HelloWorld`
```

File: demisto_sdk/commands/validate/file_validator.py [hunk 10]
```diff
@@ -83,7 +83,7 @@ class FilesValidator:
 
     def __init__(self, is_backward_check=True, prev_ver=None, use_git=False, is_circle=False,
                  print_ignored_files=False, validate_conf_json=True, validate_id_set=False, file_path=None,
-                 validate_all=False, configuration=Configuration()):
+                 validate_all=False, pack_path=False, configuration=Configuration()):
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:86]: revert changes here

File: demisto_sdk/commands/validate/file_validator.py [hunk 11]
```diff
@@ -101,6 +101,7 @@ def __init__(self, is_backward_check=True, prev_ver=None, use_git=False, is_circ
         self.validate_conf_json = validate_conf_json
         self.validate_id_set = validate_id_set
         self.file_path = file_path
+        # self.pack_path = pack_path
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:104]: remove comment

File: demisto_sdk/commands/validate/file_validator.py [hunk 12]
```diff
@@ -729,9 +747,12 @@ def is_valid_structure(self):
...
         else:
             if self.file_path:
-                print('Not using git, validating file: {}'.format(self.file_path))
-                self.is_backward_check = False  # if not using git, no need for BC checks
-                self.validate_added_files({self.file_path}, file_type=find_type(self.file_path))
+                if os.path.isfile(self.file_path):
+                    print('Not using git, validating file: {}'.format(self.file_path))
+                    self.is_backward_check = False  # if not using git, no need for BC checks
+                    self.validate_added_files({self.file_path}, file_type=find_type(self.file_path))
+                if os.path.isdir(self.file_path):
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:757]:
```suggestion
                elif os.path.isdir(self.file_path):
```

File: demisto_sdk/commands/validate/file_validator.py [hunk 13]
```diff
@@ -650,6 +651,23 @@ def validate_all_files(self):
         for file in all_files_to_validate:
             self.run_all_validations_on_file(file, file_type=find_type(file))
 
+    def validate_pack(self):
+        """Validate files in a specified pack"""
+        print_color(f'Validating {self.file_path}', LOG_COLORS.GREEN)
+        packs = glob(f'{PACKS_DIR}/*')
+        if self.file_path not in packs:
+            raise Exception(F'File {self.file_path} was not found\nMake sure you are using relative path')
```
Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:659]:
```suggestion
            raise Exception(f'Pack {self.file_path} was not found\nMake sure you are using relative path')
```

Comment (Itay4) [inline demisto_sdk/commands/validate/file_validator.py:659]: add an example here for a relative path

File: demisto_sdk/commands/validate/file_validator.py (hunk 6 above)
Comment (Itay4) [reply demisto_sdk/commands/validate/file_validator.py:667]: note my comment here

File: demisto_sdk/__main__.py [hunk 14]
```diff
@@ -207,27 +207,30 @@ def unify(**kwargs):
...
 @click.option(
-    '-p', '--path', help='Path of file to validate specifically, outside of a git directory.'
+    '-p', '--path', help='Path of file to validate specifically, outside of a git directory.', hidden=True
 )
 @click.option(
     '-a', '--validate-all', is_flag=True, show_default=True, default=False,
     help='Whether to run all validation on all files or not'
 )
+@click.option(
+    '-i', '--input', help='The path of a pack to validate specifically.'
```
Comment (Itay4) [inline demisto_sdk/__main__.py:217]:
```suggestion
    '-i', '--input', help='The path of the content pack/file to validate specifically.'
```

File: demisto_sdk/commands/common/tests/tools_test.py [hunk 15]
```diff
@@ -125,6 +125,10 @@ def test_get_remote_md_file_origin(self):
         hello_world_readme = tools.get_remote_file('Packs/HelloWorld/README.md', 'master')
         assert hello_world_readme == {}
 
+    def test_is_test_file(self):
+        test_file = tools.is_test_file('Packs/HelloWorld/Integrations/HelloWorld/search_alerts.json ')
```
Comment (Itay4) [inline demisto_sdk/commands/common/tests/tools_test.py:129]:
```suggestion
        test_file = tools.is_test_file('Packs/HelloWorld/Integrations/HelloWorld/search_alerts.json')
```

File: demisto_sdk/commands/common/tests/tools_test.py [hunk 16]
```diff
@@ -125,6 +125,10 @@ def test_get_remote_md_file_origin(self):
...
+        assert not test_file
```
Comment (Itay4) [inline demisto_sdk/commands/common/tests/tools_test.py:130]: lets also add a case for a test file

File: demisto_sdk/commands/common/tools.py [hunk 17]
```diff
@@ -731,3 +731,8 @@ def is_file_from_content_repo(file_path: str) -> Tuple[bool, str]:
         return True, '/'.join(input_path_parts[len(content_path_parts):])
     else:
         return False, ''
+
+
+def is_test_file(file_):
```
Comment (Itay4) [inline demisto_sdk/commands/common/tools.py:736]:
```suggestion
def is_test_file(file_: str) -> bool:
```

Comment (Itay4) [inline demisto_sdk/commands/common/tools.py:736]: add docstring

File: demisto_sdk/commands/common/tools.py [hunk 18]
```diff
@@ -731,3 +731,8 @@ def is_file_from_content_repo(file_path: str) -> Tuple[bool, str]:
...
+    if any(test_file in file_.lower() for test_file in TESTS_DIRECTORIES):
+        return True
```
Comment (Itay4) [inline demisto_sdk/commands/common/tools.py:738]: need to `return False` if not

File: demisto_sdk/commands/common/tools.py [hunk 19]
```diff
@@ -731,3 +732,11 @@ def is_file_from_content_repo(file_path: str) -> Tuple[bool, str]:
         return True, '/'.join(input_path_parts[len(content_path_parts):])
     else:
         return False, ''
+
+
+def is_test_file(file_: str):
```
Comment (Itay4) [inline demisto_sdk/commands/common/tools.py:737]:
```suggestion
def is_test_file(file_: str) -> bool:
```

File: demisto_sdk/commands/common/tools.py [hunk 20]
```diff
@@ -731,3 +732,11 @@ def is_file_from_content_repo(file_path: str) -> Tuple[bool, str]:
...
+    """Check if the file is a test file under testdata, test_data or data_test"""
+    if any(test_file in file_.lower() for test_file in TESTS_DIRECTORIES):
+        return True
+    else:
+        False
```
Comment (Itay4) [inline demisto_sdk/commands/common/tools.py:742]:
```suggestion
        return False
```

File: demisto_sdk/commands/common/tools.py [hunk 21]
```diff
@@ -731,3 +732,11 @@ def is_file_from_content_repo(file_path: str) -> Tuple[bool, str]:
         return True, '/'.join(input_path_parts[len(content_path_parts):])
     else:
         return False, ''
+
+
+def is_test_file(file_: str):
+    """Check if the file is a test file under testdata, test_data or data_test"""
```
Comment (Itay4) [inline demisto_sdk/commands/common/tools.py:738]: add args and returns to docstring

File: demisto_sdk/commands/validate/tests/validators_test.py [hunk 22]
```diff
@@ -432,3 +434,14 @@ def test_script_valid_rn(self, mocker, file_path, file_type):
         file_validator = FilesValidator(validate_conf_json=False)
         file_validator.validate_added_files(file_path, file_type)
         assert file_validator._is_valid
+
+
+TEST_FILES_PATH = join(git_path(), "demisto_sdk/tests/test_files/content_repo_example/")
```
Comment (Itay4) [inline demisto_sdk/commands/validate/tests/validators_test.py:439]: why did you add this ?

File: demisto_sdk/tests/integration_tests/pack_validation_test.py [hunk 23]
```diff
@@ -0,0 +1,36 @@
+import os
```
Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:1]: no need for new file, move it to `demisto-sdk/demisto_sdk/tests/integration_tests/validate_integration_test.py`

File: demisto_sdk/tests/integration_tests/pack_validation_test.py [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+from demisto_sdk.commands.common.git_tools import git_path
+
+VALIDATE_CMD = "validate"
+TEST_FILES_PATH = join(git_path(), "demisto_sdk/tests/test_files/content_repo_example/")
+AZURE_FEED_PACK_PATH = "Packs/FeedAzure"
+AZURE_FEED_INVALID_PACK_PATH = join(TEST_FILES_PATH, "Packs/FeedAzure")
+
+
+class TestPack:
+    def test_validate_pack(self):
```
Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:15]:
```suggestion
    def test_integration_validate_pack_positive(self):
```

Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:15]: add docstring `Given-When-Then`

File: demisto_sdk/tests/integration_tests/pack_validation_test.py [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+        os.chdir(TEST_FILES_PATH)
```
Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:16]: why is this needed?

File: demisto_sdk/tests/integration_tests/pack_validation_test.py [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+        runner = CliRunner(mix_stderr=False)
+        result = runner.invoke(main, [VALIDATE_CMD, "-i", AZURE_FEED_PACK_PATH])
+        assert "Starting validating files structure" in result.output
+        assert f"Validating Packs/FeedAzure" in result.output
+        assert f'Validating Packs/FeedAzure/Integrations/FeedAzure/FeedAzure.yml' in result.output
+        assert f'Validating Packs/FeedAzure/IncidentFields/incidentfield-city.json' in result.output
+        assert f'Validating Packs/FeedAzure/TestPlaybooks/FeedAzure_test.yml' in result.output
+        assert f'Validating Packs/FeedAzure/TestPlaybooks/script-prefixed_automation.yml' in result.output
```
Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:24]: add `assert "The files are valid" in result.stdout`

Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:24]: add `assert result.stderr == ""`

File: demisto_sdk/tests/integration_tests/pack_validation_test.py [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+
+    def test_invalid_pack_path(self):
+        runner = CliRunner(mix_stderr=False)
+        result = runner.invoke(main, [VALIDATE_CMD, "-i", 'content_repo_example/Packs/FeedAzure'])
+        assert result.exit_code == 1
+        assert 'content_repo_example/Packs/FeedAzure was not found' in result.output
```
Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:30]: add `assert result.stderr == ""`

File: demisto_sdk/tests/integration_tests/pack_validation_test.py [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+        runner = CliRunner(mix_stderr=False)
+        result = runner.invoke(main, [VALIDATE_CMD, "-i", AZURE_FEED_PACK_PATH])
+        assert "Starting validating files structure" in result.output
+        assert f"Validating Packs/FeedAzure" in result.output
+        assert f'Validating Packs/FeedAzure/Integrations/FeedAzure/FeedAzure.yml' in result.output
+        assert f'Validating Packs/FeedAzure/IncidentFields/incidentfield-city.json' in result.output
+        assert f'Validating Packs/FeedAzure/TestPlaybooks/FeedAzure_test.yml' in result.output
+        assert f'Validating Packs/FeedAzure/TestPlaybooks/script-prefixed_automation.yml' in result.output
+
+    def test_invalid_pack_path(self):
```
Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:26]:
```suggestion
    def test_integration_validate_invalid_pack_path(self):
```

Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:26]: add docstring `Given-When-Then`

File: demisto_sdk/tests/integration_tests/pack_validation_test.py [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+        runner = CliRunner(mix_stderr=False)
+        result = runner.invoke(main, [VALIDATE_CMD, "-i", 'content_repo_example/Packs/FeedAzure'])
+        assert result.exit_code == 1
+        assert 'content_repo_example/Packs/FeedAzure was not found' in result.output
+
+    def test_pack_metadata(self):
+        os.chdir(TEST_FILES_PATH)
```
Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:33]: why is this needed?

File: demisto_sdk/tests/integration_tests/pack_validation_test.py [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+        assert f'Validating Packs/FeedAzure/TestPlaybooks/FeedAzure_test.yml' in result.output
+        assert f'Validating Packs/FeedAzure/TestPlaybooks/script-prefixed_automation.yml' in result.output
+
+    def test_invalid_pack_path(self):
+        runner = CliRunner(mix_stderr=False)
+        result = runner.invoke(main, [VALIDATE_CMD, "-i", 'content_repo_example/Packs/FeedAzure'])
+        assert result.exit_code == 1
+        assert 'content_repo_example/Packs/FeedAzure was not found' in result.output
+
+    def test_pack_metadata(self):
```
Comment (Itay4) [inline demisto_sdk/tests/integration_tests/pack_validation_test.py:32]: what are we testing here?

File: demisto_sdk/tests/integration_tests/pack_validation_test.py (hunk 26 above)
Comment (Shellyber) [reply demisto_sdk/tests/integration_tests/pack_validation_test.py:24]: They are not.
Usually there is something missing like `propagation labels` and things like that.

File: demisto_sdk/tests/integration_tests/pack_validation_test.py (hunk 30 above)
Comment (Shellyber) [reply demisto_sdk/tests/integration_tests/pack_validation_test.py:32]: That the function is testing the pack-metadata-file

File: demisto_sdk/tests/integration_tests/pack_validation_test.py (hunk 26 above)
Comment (Itay4) [reply demisto_sdk/tests/integration_tests/pack_validation_test.py:24]: ok so this is a negative test, and not positive 
lets assert the non-valid errors and lets add another positive test where all files are valid

File: demisto_sdk/__main__.py [hunk 31]
```diff
@@ -207,27 +207,30 @@ def unify(**kwargs):
     'If the --post-commit flag is not supplied: validation will run on all changed files in the current branch, '
     'both committed and not committed. ')
 @click.option(
-    '-p', '--path', help='Path of file to validate specifically, outside of a git directory.'
+    '-p', '--path', help='Path of file to validate specifically, outside of a git directory.', hidden=True
```
Comment (jochman) [inline demisto_sdk/__main__.py:210]: can't we just remove this? we're not breaking any bc before 1.0

Comment (Shellyber) [reply demisto_sdk/__main__.py:210]: No, Itay said to leave it. It's hidden anyway
